pandas/core/groupby/groupby.py

```diff
@@ -2988,13 +2988,15 @@ def _reindex_output(
             index, _ = MultiIndex.from_product(
                 levels_list, names=self.grouper.names
             ).sortlevel()
-        except ValueError as err:
-            raise ValueError(
-                "Product space too large to allocate arrays! You're trying to "
-                "group over too complex a column cross-product to fit in memory. "
-                "If any of those columns are categorical, you may be able "
-                "to circumvent this issue by passing `observed=True`."
-            ) from err
+        except (ValueError, MemoryError) as err:
```

Comment:
The memory error is thrown from within NumPy, with the message `Unable to allocate <number> <byte unit> for an array with shape <dims> and data type <dtype>`. By the time we've reached this point, we know we're dealing with a DataFrame with at least one categorical column in the grouped-by columns, and that `observed=False` so memory allocation will be larger with `observed=False` than `observed=True`.

We cannot know if the system is overloaded or not, but we can be certain that if the user passed `observed=True`, it would use less memory. I can add " or increase available memory." to the error message to a cover this case.

Reply:
> memory allocation will be larger with observed=False than observed=True

Agreed this is likely to be true but not always. However, in the case it's a reasonable size (whatever that is) and the user is out of memory we don't want to be pointing them down the wrong path. I think we also don't want to be intercepting a MemoryError and replacing it with a ValueError.

Your recommend addition makes sense to me if raise with a MemoryError. I recommend having this PR focus on the "Product space too large" ValueError. Another PR to make the MemoryError more informative would be most welcome, but better to separate.